src/mmu/mod.rs

```
@@ -0,0 +1,173 @@
+//! 
+//! Different regions of memory are treated in different ways by different pieces of hardware on
+//! the Gameboy. For example, the CPU is restricted from accessing VRAM and OAM during certain
+//! timing windows. Certain registers, such as DIV and TIMA, incur side effects when written to.
+ 
+pub mod memmap;
+mod readwrite;
+mod dma;
+mod timers;
+
+use dma::Dma;
+use memmap::*;
+use std::{cell::RefCell, rc::Rc};
+use timers::Timers;
+
+use crate::util::set_bit;
+
+pub struct Mmu {
+    dma: Dma,
+    timers: Timers,
+    rom_bank_00: [u8; ROM_BANK_0_SIZE],
+    rom_bank_01: [u8; ROM_BANK_1_SIZE],
+    vram: [u8; VRAM_SIZE],
+    exram: [u8; EXRAM_SIZE],
+    wram_0: [u8; WRAM_0_SIZE],
+    wram_1: [u8; WRAM_1_SIZE],
+    oam: [u8; OAM_SIZE],
+    restricted_memory: [u8; RESTRICTED_MEM_SIZE],
+    io: [u8; IO_SIZE],
+    hram: [u8; HRAM_SIZE],
+    ie: u8,
+
+    pub vram_lock: bool,
+    pub oam_lock: bool,
+}
+
+impl Mmu {
+    //! I chose to use the Rc RefCell for the MMU so that the CPU and PPU could borrow a mutable
+    //! reference to access it whenever they need to. It could just as easily be a global variable.
+    pub fn new() -> Rc<RefCell<Mmu>> {
+        let mmu = Mmu {
+            dma: Dma::new(),
+            timers: Timers::new(),
+            rom_bank_00: [0; ROM_BANK_0_SIZE],
+            rom_bank_01: [0; ROM_BANK_1_SIZE],
+            vram: [0; VRAM_SIZE],
+            exram: [0; EXRAM_SIZE],
+            wram_0: [0; WRAM_0_SIZE],
+            wram_1: [0; WRAM_1_SIZE],
+            oam: [0; OAM_SIZE],
+            restricted_memory: [0; RESTRICTED_MEM_SIZE],
+            io: [0; IO_SIZE],
+            hram: [0; HRAM_SIZE],
+            ie: 0,
+
+            vram_lock: false,
+            oam_lock: false,
+        };
+
+        Rc::new(RefCell::new(mmu))
+    }
+
+    // todo! Support bigger ROMs
+    pub fn load_rom(&mut self, path: &str) -> bool {
```

This is a great opportunity to use Error handling.   Instead of swallowing the error below, you could declare this to return `std::result:;Result((), Error)` and then below try:

```
let from = match std::fs::read(path)?;
```
and be done with it.